File: src/audio/protocol.ts
export const AudioMessage = {
  Playback: 0x01,
  ClearPlayback: 0x02,
  SetMuted: 0x03,
  Shutdown: 0x04,
  Capture: 0x10,
  Levels: 0x11,
  Ready: 0x12,
  Error: 0x13,
} as const;

export interface AudioFrame { type: number; payload: Buffer }

export function encodeAudioFrame(type: number, payload: Buffer = Buffer.alloc(0)): Buffer {
  const frame = Buffer.allocUnsafe(5 + payload.length);
  frame[0] = type;
  frame.writeUInt32LE(payload.length, 1);
  payload.copy(frame, 5);
  return frame;
}

export class AudioFrameDecoder {
  private pending = Buffer.alloc(0);
  push(chunk: Buffer): AudioFrame[] {
    if (!chunk.length) return [];
    this.pending = this.pending.length ? Buffer.concat([this.pending, chunk]) : chunk;
    const frames: AudioFrame[] = [];
    let offset = 0;
    while (this.pending.length - offset >= 5) {
      const size = this.pending.readUInt32LE(offset + 1);
      if (size > 8 * 1024 * 1024) throw new Error(`audio helper frame too large: ${size}`);
      if (this.pending.length - offset < 5 + size) break;
      const type = this.pending[offset]!;
      frames.push({ type, payload: this.pending.subarray(offset + 5, offset + 5 + size) });
      offset += 5 + size;
    }
    if (offset) this.pending = this.pending.subarray(offset);
    return frames;
  }
}

File: audio-helper/internal/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"
)

type Type byte

const (
	Playback      Type = 0x01
	ClearPlayback Type = 0x02
	SetMuted      Type = 0x03
	Shutdown      Type = 0x04

	Capture Type = 0x10
	Levels  Type = 0x11
	Ready   Type = 0x12
	Error   Type = 0x13
)

const MaxPayload = 8 << 20

type Frame struct {
	Type    Type
	Payload []byte
}

func Read(r io.Reader) (Frame, error) {
	var header [5]byte
	if _, err := io.ReadFull(r, header[:]); err != nil {
		return Frame{}, err
	}
	n := binary.LittleEndian.Uint32(header[1:])
	if n > MaxPayload {
		return Frame{}, fmt.Errorf("protocol payload too large: %d", n)
	}
	payload := make([]byte, int(n))
	if n > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return Frame{}, err
		}
	}
	return Frame{Type: Type(header[0]), Payload: payload}, nil
}

func Write(w io.Writer, typ Type, payload []byte) error {
	if len(payload) > MaxPayload {
		return fmt.Errorf("protocol payload too large: %d", len(payload))
	}
	var header [5]byte
	header[0] = byte(typ)
	binary.LittleEndian.PutUint32(header[1:], uint32(len(payload)))
	if _, err := w.Write(header[:]); err != nil {
		return err
	}
	if len(payload) > 0 {
		_, err := w.Write(payload)
		return err
	}
	return nil
}
